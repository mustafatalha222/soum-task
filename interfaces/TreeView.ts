export interface ITreeNode {
  name: string;
  children?: ITreeNode[];
  checked?: boolean;
  id: string;
  description?: string;
}

export interface ITreeViewProps {
  data: ITreeNode[];
  onSelectionChange?: (selectedNodes: ITreeNode[]) => void;
}
