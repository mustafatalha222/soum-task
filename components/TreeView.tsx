import { ITreeNode, ITreeViewProps } from "@/interfaces/TreeView";
import Checkbox from "expo-checkbox";
import React, { memo } from "react";
import { View, Pressable, StyleSheet, FlatList } from "react-native";
import { Colors } from "@/constants/Colors";
import { useExpandedNodes } from "@/hooks/useExpandedNodes";
import { useCheckedState } from "@/hooks/useCheckedState";
import { ThemedText } from "./ThemedText";

const TreeView = ({ data, onSelectionChange }: ITreeViewProps) => {
  const { nodes, toggleNodeCheckbox } = useCheckedState(
    data,
    onSelectionChange
  );
  const { expandedNodes, handleToggleNode } = useExpandedNodes();

  const renderItem = ({ item }: { item: ITreeNode }) => {
    const isExpanded = expandedNodes.includes(item.id);
    return (
      <View style={styles.nodeContainer}>
        <View style={styles.row}>
          <Checkbox
            testID={`checkbox-${item.id}`}
            value={item.checked}
            onValueChange={() => toggleNodeCheckbox(item)}
          />
          <Pressable onPress={() => handleToggleNode(item.id)}>
            <View style={styles.ThemedTextContainer}>
              <ThemedText type={item.children ? "defaultSemiBold" : "default"}>
                {item.name}
              </ThemedText>
              {item.description && (
                <ThemedText style={styles.nodeDescription}>
                  {item.description}
                </ThemedText>
              )}
            </View>
          </Pressable>
        </View>
        {isExpanded && item.children && (
          <FlatList
            data={item.children}
            renderItem={renderItem}
            keyExtractor={(child) => child.id}
          />
        )}
      </View>
    );
  };

  if (!data.length)
    return (
      <ThemedText type="subtitle" style={styles.noData}>
        No Data Found
      </ThemedText>
    );
  return (
    <FlatList
      data={nodes}
      renderItem={renderItem}
      keyExtractor={(node) => node.id}
    />
  );
};

const styles = StyleSheet.create({
  nodeContainer: {
    marginLeft: 20,
  },
  row: {
    flexDirection: "row",
    alignItems: "center",
    backgroundColor: Colors.light.grey,
    padding: 4,
    borderRadius: 8,
    marginBottom: 8,
    gap: 1,
  },
  ThemedTextContainer: {
    flex: 1,
    marginLeft: 10,
  },
  nodeDescription: {
    color: "#777",
    fontSize: 12,
    lineHeight: 12,
  },
  noData: {
    margin: 20,
    textAlign: "center",
  },
});

export default memo(TreeView);
