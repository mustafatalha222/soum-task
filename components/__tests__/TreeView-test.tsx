import * as React from "react";
import renderer from "react-test-renderer";
import TreeView from "../TreeView";
import { render, fireEvent, RenderAPI } from "@testing-library/react-native";

const sampleData = [
  {
    id: "1",
    name: "mobile",
    description: "Mobile Devices",
    children: [
      { id: "2", name: "Apple" },
      { id: "3", name: "Samsung" },
    ],
  },
];

const firstChild = sampleData[0];

describe("TreeView Cases with Data", () => {
  let component: RenderAPI;

  beforeEach(() => {
    component = render(<TreeView data={sampleData} />);
  });

  it("renders parent correctly with name and description", () => {
    const { getByText } = component;
    expect(getByText(firstChild.name)).toBeTruthy();
    expect(getByText(firstChild.description)).toBeTruthy();
  });

  it("expands nodes correctly", () => {
    const { getByText } = component;
    const node = getByText(firstChild.name);
    fireEvent.press(node);
    expect(getByText(firstChild!.children![0].name)).toBeTruthy();
  });

  it("collapses nodes correctly when press 2", () => {
    const { getByText, queryByText } = component;
    const node = getByText(firstChild.name);
    // Pressed twice to collapse
    fireEvent.press(node);
    fireEvent.press(node);
    expect(queryByText(firstChild!.children![0].name)).toBeFalsy();
  });

  it("checkbox toggle check", () => {
    const { getByTestId } = component;
    const checkbox = getByTestId(`checkbox-${firstChild.id}`);
    expect(checkbox).toBeDefined();

    fireEvent.press(checkbox);
    expect(checkbox.props.accessibilityState.checked).toBe(true);

    fireEvent.press(checkbox);
    expect(checkbox.props.accessibilityState.checked).toBe(false);
  });

  it("should show all children when parent text is clicked with false checkbox", () => {
    const { getByTestId, getByText } = component;
    const node = getByText(firstChild.name);
    fireEvent.press(node);

    firstChild.children.forEach((checkbox) => {
      const childCheckbox = getByTestId(`checkbox-${checkbox.id}`);
      expect(childCheckbox).toBeDefined();
      expect(childCheckbox.props.accessibilityState.checked).toBe(false);
    });
  });

  it("should check all children when parent is checked", () => {
    const { getByTestId, getByText } = component;
    const node = getByText(firstChild.name);
    fireEvent.press(node);
    const parentCheckbox = getByTestId(`checkbox-${firstChild.id}`);
    fireEvent.press(parentCheckbox);

    firstChild.children.forEach((checkbox) => {
      const childCheckbox = getByTestId(`checkbox-${checkbox.id}`);
      expect(childCheckbox).toBeDefined();
      expect(childCheckbox.props.accessibilityState.checked).toBe(true);
    });
  });
});

describe("TreeView Cases without Data", () => {
  it("No data case", () => {
    const { queryByText } = render(<TreeView data={[]} />);
    expect(queryByText(/no data/i)).toBeTruthy();
  });
});

describe("Tree View Snapshot", () => {
  it(`snapshot test for TreeView correctly`, () => {
    const tree = renderer.create(<TreeView data={sampleData} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
