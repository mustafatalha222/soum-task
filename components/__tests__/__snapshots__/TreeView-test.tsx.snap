// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree View Snapshot snapshot test for TreeView correctly 1`] = `
<RCTScrollView
  data={
    [
      {
        "checked": false,
        "children": [
          {
            "checked": false,
            "children": undefined,
            "id": "2",
            "name": "Apple",
          },
          {
            "checked": false,
            "children": undefined,
            "id": "3",
            "name": "Samsung",
          },
        ],
        "description": "Mobile Devices",
        "id": "1",
        "name": "mobile",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={[]}
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "flexDirection": "column",
            "marginLeft": 20,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#f0f0f0",
              "borderRadius": 8,
              "flexDirection": "row",
              "gap": 1,
              "marginBottom": 8,
              "padding": 4,
            }
          }
        >
          <View
            accessibilityRole="checkbox"
            accessibilityState={
              {
                "busy": undefined,
                "checked": false,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                {
                  "borderColor": "#657786",
                  "borderRadius": 2,
                  "borderWidth": 2,
                  "height": 20,
                  "width": 20,
                },
                undefined,
                false,
                false,
                undefined,
                false,
              ]
            }
            testID="checkbox-1"
          />
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <View
              style={
                {
                  "flex": 1,
                  "marginLeft": 10,
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "color": "#11181C",
                    },
                    undefined,
                    undefined,
                    {
                      "fontSize": 16,
                      "fontWeight": "600",
                      "lineHeight": 24,
                    },
                    undefined,
                    undefined,
                    undefined,
                  ]
                }
              >
                mobile
              </Text>
              <Text
                style={
                  [
                    {
                      "color": "#11181C",
                    },
                    {
                      "fontSize": 16,
                      "lineHeight": 24,
                    },
                    undefined,
                    undefined,
                    undefined,
                    undefined,
                    {
                      "color": "#777",
                      "fontSize": 12,
                      "lineHeight": 12,
                    },
                  ]
                }
              >
                Mobile Devices
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
